### ISSUE TITLE
challenges need a "Report" button to flag violations of the Community Guidelines

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#5557]
_Edit by Alys 2021-09-15_: **There is a mostly-complete PR at https://github.com/HabitRPG/habitica/pull/11920 which another Blacksmith could finish to resolve this issue.**

_In this issue "mod" means moderator or staff member. In the codebase, they can be identified by `user.contributor.admin` being set to `true`._

Players sometimes create Challenges that contain violations of the Community Guidelines. There needs to be a "Report" button for other players to flag such Challenges as needing attention from mods.

The "Report" button should appear when:
- seeing the Challenge in the My Challenges page
- seeing the Challenge in the Discover Challenges page
- viewing the individual Challenge's page (i.e., pages with links like this: habitica.com/challenges/2def5ee7-291b-4ec0-860e-33c2d3d09551)

The button should work in much the same way as it does for chat messages. When you click the button, a confirmation modal appears with two buttons, "Cancel" and "Report". If you click "Cancel", the modal closes and no flag is made. When you click "Report", a report is made (see below for details). When mods view that modal for a Challenge that has already been flagged, they see an additional button, "Reset Flag Count" (see below for details).

The screenshot at the bottom of this post shows the modal a mod sees when they click the "Report" button on a chat message. The chat report modal that a non-mod sees is identical except that the "Reset Flag Count" button is not there.
The modal for reporting a Challenge should look the same except:
- The text of the post ("A test post for flagging." in the screenshot) is replaced with the Challenge's name.
- The "Why are you reporting this post?" subheading should have every "post" changed to "Challenge".
- The "Please help our moderators..." sentence should have "post" changed to "Challenge".
- The "Are you sure..." paragraph should have every instance of "post" changed to "Challenge".
- The "Reset Flag Count" button should appear for only mods.

When the "Report" button is clicked, the following actions need to occur:
- an email is sent to the same recipients who receive notifications about flagged chat messages
- a message is sent to the moderator Slack channel, as is done for flagged chat messages
- the Challenge's document in the database is modified to contain details like these:
```
"flags": {
	"d904bd62-da08-416b-a816-ba797c9ee265": true,
	"3e595299-3d8a-4a10-bfe0-88f555e4aa0c": true
},
"flagCount": 2,
```
Those UUIDs are the User IDs (not Usernames!) of the people who have clicked the "Report" button. The `flagCount` increases by 1 for each non-mod who clicks the button. If a mod clicks the button, the flagCount is immediately set to 5. Those details are the same as for chat messages, so the contributor who works on this can copy what's in the chat schema and code.
When a Challenge has never been flagged, that data looks like below (again, this is the same as the [default data for chat messages](https://github.com/HabitRPG/habitica/blob/d9830950aacc3343d708eccd1a2a33c585e797de/website/server/models/group.js#L450-L451)):
```
"flags": {},
"flagCount": 0,
```

When a mod clicks the "Reset Flag Count" button, these things must happen (they are all the same as is done when resetting flags on a chat message):
- an email is sent to the same recipients who receive notifications about un-flagged chat messages
- the Challenge's `flagCount` is set to `0`
- if the mod's User ID is currently recorded as `true` in the `flags` object, then it's changed to `false`
- _no other changes are made to the `flags` object_
- _no Slack notification is made for unflagging_

When a Challenge has a `flagCount` greater than zero, mods should see a message on the Challenge saying that it's been flagged (e.g., just the word "Flagged"). That message will never need to be seen by non-mod players so it doesn't really matter what it looks like as long as it's clearly visible (e.g., close to the challenge's name). I suggest red text, as is done for a similar message for mods on [flagged chat messages](https://github.com/HabitRPG/habitica/blob/d9830950aacc3343d708eccd1a2a33c585e797de/website/client/components/chat/chatMessages.vue#L23-L24). I'm happy for the contributor who works on this to make a decision about it and show us a screenshot in your PR. It's not worth spending time designing that carefully. The message should appear in these three places:
- seeing the Challenge in the My Challenges page
- seeing the Challenge in the Discover Challenges page
- viewing the individual Challenge's page (i.e., pages with links like this: habitica.com/challenges/2def5ee7-291b-4ec0-860e-33c2d3d09551)

For the first version of this feature, there is no need for flagged Challenges to be hidden from any players, however that is a desired feature and can be added later. See https://github.com/HabitRPG/habitica/issues/9725 for details. For simplicity and to get this feature done more quickly, I suggest that both issues are not worked on at the same time. ~_Previous text: If the contributor who wants to work on this feature also wants to work on that issue at the same time, that's fine, but this first issue is more important and the details of hiding challenges are still being sorted out so that other issue isn't currently ready to be actioned._~

On a local install, the contributor who works on this can turn one of their test accounts into an admin account by using the "Make Admin" button in the Debug menu in the footer of the website, then reload the website. Comment in this issue if that doesn't seem to do what you need - it's been a while since I used that Debug feature. :)

![image](https://user-images.githubusercontent.com/1495809/48295329-a2302a80-e4d6-11e8-81c7-cb8b7333a25c.png)

**Comment by Alys** [comment]
If anyone wants to experiment with how the chat message flag system works, I (or other admins) can make some test posts in the Aspiring Coders guild that you can flag. Those of us who receive email notifications of flags won't mind; the test messages can make it clear that they are for testing flagging. Just ask here if you want that.

**Comment by lemoness** [comment]
_edit by Alys 2017-12-17: ignore this comment. This requirement is being considered in https://github.com/HabitRPG/habitica/issues/9725_

Text for the owner of a flagged Challenge:

"This Challenge has been hidden by members of the community because it violates [the Community Guidelines](link). Staff members or moderators will contact you shortly via email with instructions for how to restore your Challenge! If over 48 hours have passed and you have not heard from them, please email leslie@habitrpg.com for assistance."

**Comment by SabreCat** [comment]
This will need enough UI implementation that it should be placed on hold until the frontend redesign is ready.

**Comment by lemoness** [comment]
We are now welcoming help on this issue, but you'll need to post here to finalize the design before implementing it!

**Comment by MathWhiz** [comment]
Should this be labeled with `status: issue: on hold - needs designs`?

**Comment by Alys** [comment]
This is now ready for a contributor to start working on it.

However @Tressley will want to decide on the exact position of the "Report" icon in the three locations listed in the top post (My Challenges, Discover Challenges, individual challenge pages). Until that's done, the contributor who works on this issue could add it in any position to help them develop the other parts of the fix for this issue.

I've deleted some of the old comments, which were about how the issue should be on hold for the website redesign.

**Comment by TheHollidayInn** [comment]
This take this as it is on my list of mod tools :D

**Comment by Alys** [comment]
I've updated the top post in this issue, deleted a now-irrelevant comment about previous work done on it, and marked it as help wanted.

**Comment by alecbrick** [comment]
I'm down to take a look at this. I'll start working from #10230.

**Comment by shanaqui** [comment]
@alecbrick Hi! Are you still interested in working on this? If so, just let us know!

**Comment by shanaqui** [comment]
Hi @alecbrick! I've marked this as help wanted again, since you didn't reply, but if you're still interested then feel free to pick it back up.

**Comment by alecbrick** [comment]
Sounds good. Sorry for not getting back to this earlier - I've been a bit busy with work, but I'll try and get back to this when I find the time.

**Comment by codexdelta** [comment]
Hi, is this issue available for me? 😃

**Comment by Alys** [comment]
@codexdelta Sure, thank you! Please go ahead. Post here if you have questions. :)

**Comment by alecbrick** [comment]
@codexdelta Hi! Are you starting from scratch, or working off of [my existing PR](https://github.com/HabitRPG/habitica/pull/11920)?

**Comment by alecbrick** [comment]
I might close my existing PR (since I'm no longer working on it), but feel free to reference and copy anything from it @codexdelta.

**Comment by alecbrick** [comment]
Just kidding I accidentally finished it up instead

**Comment by SabreCat** [comment]
Opening back up as alecbrick has stepped away from the work. It may not require much more effort--could be a quick win for another contributor! 

_EDIT by Alys 2021-09-15:_ alecbrick's PR is https://github.com/HabitRPG/habitica/pull/11920

**Comment by fmaximus** [comment]
@SabreCat seems the PR of @alecbrick got merged. "help wanted" not necessary I guess.

**Comment by Alys** [comment]
@fmaximus Thanks for posting about that!

I think there's actually been some confusion with the links and this issue doesn't yet have a PR that resolves it, as far as I can see. The link in the post here from 18 days ago is for a different issue.

@alecbrick Do you still have a PR open that you're working on? Or should this stay open for someone else to take up?

**Comment by alecbrick** [comment]
Yup! The PR can be found [here](https://github.com/HabitRPG/habitica/pull/11920). I'm no longer working on it, but it's almost complete.

**Comment by Alys** [comment]
@alecbrick Thanks for replying! I've edited the comment from a few weeks back to insert that link, and also added it to the top post. I've made similar changes in https://github.com/HabitRPG/habitica/issues/9725

@fmaximus were you interested in working on this issue?

**Comment by matchajune** [comment]
I would like to take this one as I am going to work on #9725

**Comment by CuriousMagpie** [comment]
Sure thing, @matchajune Thanks!

**Comment by matchajune** [comment]
How should the "Report" button be shown for these two locations:

- seeing the Challenge in the My Challenges page
- seeing the Challenge in the Discover Challenges page

**Comment by matchajune** [comment]
<img width="711" alt="Screenshot 2023-03-06 at 8 51 41 PM" src="https://user-images.githubusercontent.com/6742066/223328626-c2fc3e74-d7ae-4c8e-b932-0ef885b82cd5.png">
<img width="1014" alt="Screenshot 2023-03-06 at 9 26 27 PM" src="https://user-images.githubusercontent.com/6742066/223328617-df892aa3-79a4-4ba7-8a69-828f90781793.png">

**Comment by matchajune** [comment]
accidentally removed myself as an assignee; can assign me again.

**Comment by tdmiller1** [comment]
Should this issue be marked as closed? Since it's been fixed with: https://github.com/HabitRPG/habitica/pull/14756

**Comment by Tressley** [comment]
Thanks for the ping @tdmiller1. This was fixed in #14756.
